sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+
+	// Unpack the base structure.
+	if err := msg.MsgHeader.UnmarshalBinary(data); err != nil {
+		return err
+	}
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	var raw msgNotificationFilterKernel
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, &raw); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	// Unpack variable length elements.
+	unpacker := newStringUnpacker(data)
+	ns, err := unpacker.UnpackString(raw.NS)
+	if err != nil {
+		return fmt.Errorf("%s: cannot unpack namespace: %v", prefix, err)
+	}
+	filter, err := unpacker.UnpackString(raw.Filter)
+	if err != nil {
+		return fmt.Errorf("%s: cannot unpack filter: %v", prefix, err)
+	}
+
+	// Put everything together.
+	msg.ModeSet = ModeSet(raw.ModeSet)
+	msg.NameSpace = ns
+	msg.Filter = filter
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotificationFilter) MarshalBinary() (data []byte, err error) {
+	var raw msgNotificationFilterKernel
+	packer := newStringPacker(raw)
+	raw.Version = 2
+	raw.ModeSet = uint32(msg.ModeSet)
+	raw.NS = packer.PackString(msg.NameSpace)
+	raw.Filter = packer.PackString(msg.Filter)
+	raw.Length = packer.TotalLen()
+	msgBuf := bytes.NewBuffer(make([]byte, 0, raw.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(msgBuf, order, raw); err != nil {
+		return nil, err
+	}
+	if _, err := msgBuf.Write(packer.Bytes()); err != nil {
+		return nil, err
+	}
+	return msgBuf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotificationFilter) Validate() error {
+	if !msg.ModeSet.IsValid() {
+		return fmt.Errorf("unsupported modeset: %d", msg.ModeSet)
+	}
+	return nil
+}
+
+// MsgNotification describes a kernel notification message.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif {
+//	  struct apparmor_notif_common base;
+//	  __u16 ntype;        /* notify type */
+//	  __u8 signalled;
+//	  __u8 flags;
+//	  __u64 id;           /* unique id, not globally unique*/
+//	  __s32 error;        /* error if unchanged */
+//	} __attribute__((packed));
+type MsgNotification struct {
+	MsgHeader
+	// NotificationType describes the kind of notification message used.
+	// Currently the kernel only sends APPARMOR_NOTIF_OP messages.
+	NotificationType NotificationType
+	// XXX: Signaled seems to be unused.
+	Signalled uint8
+	// Set to 1 to NOT cache
+	Flags uint8
+	// ID is an opaque kernel identifier of the notification message. It must be
+	// repeated in the MsgNotificationResponse if one is sent back.
+	ID uint64
+	// XXX: This seems to be unused and clashes with identical field in
+	// MsgNotificationOp.
```

Comment:
it this an issue?

Reply:
These structs are all drawn from apparmor, AFAIK. I'm going to ask @jrjohansen to make a quick pass over message.go and see if he knows what the state of a few of these XXX comments are. Also, whether `MsgNotificationUpdate` should have been removed.

Reply:
So. Part of the issue here is the kernel interface is C expressed in C code and using unions, so yes we get overlapping fields. Mapping this over to go is not the cleanest.

Eg.
```
struct __attribute__((__packed__)) apparmor_notif {
	struct apparmor_notif_common base;
	__u16 ntype;			/* notify type */
	__u8 signalled;
	__u8 reserved;
	__u64 id;			/* unique id, not gloablly unique*/
	__s32 error;			/* error if unchanged */
};

struct __attribute__((__packed__)) apparmor_notif_update {
	struct apparmor_notif base;
	__u16 ttl;			/* max keep alives left */
};

...
union __attribute__((__packed__)) apparmor_notif_all {
	struct apparmor_notif_common common;
	struct apparmor_notif_filter filter;
	struct apparmor_notif base;
	struct apparmor_notif_op op;
	struct apparmor_notif_file file;
};
```

Reply:
snapd currently does not use MsgNotifcationUpdate and it can be dropped. It can always be readded when/if needed.